software/contrib/probapoly.py

```diff
@@ -0,0 +1,242 @@
+                self.patternLengthPrevious = self.patternLength
+
+    def handleClock(self):
+        
+        # Play upper gate
+        if self.step == 1 or (self.step-1) % self.upper == 0:
+            cv1.value(1)
+
+        # Output trigger with fixed and unrelated probabilities
+            if randint(0,99) < self.upperProb1:
+                cv2.value(1)
+
+            if randint(0,99) < self.upperProb2:
+                cv3.value(1)
+
+        # Play lower gate
+        if self.step == 1 or (self.step-1) % self.lower == 0:
+            cv4.value(1)
+
+            # Output trigger with fixed and unrelated probabilities
+            if randint(0,99) < self.lowerProb1:
+                cv5.value(1)
+
+            if randint(0,99) < self.lowerProb2:
+                cv6.value(1)
+
+    # Generate pattern length by finding the lowest common multiple (LCM) and greatest common divisor (GCD)
+    # https://www.programiz.com/python-programming/examples/lcm
+    def lcm(self, x, y):
+        return (x*y)//self.computeGcd(x,y)
+
+    def computeGcd(self, x, y):
+        while(y):
+            x, y = y, x % y
+        return x
+
+    def getUpper(self):
+        # Mode 2, use the analogue input voltage to set the upper ratio value
+        if self.ainValue > 0.9 and self.ainMode == 2:
+            self.upper = int((self.maxPolyVal / 100) * self.ainValue) + 1
+        else:
+            self.upper = k1.read_position(self.maxPolyVal) + 1
+
+    def getLower(self):
+        self.lower = k2.read_position(self.maxPolyVal) + 1
+
+    def getAinValue(self):
+        self.ainValue = 100 * ain.percent()
+        #print(self.ainValue)
+
+    def updateScreen(self):
+        # Clear the screen
+        oled.fill(0)
+
+        rectLeftX = 20
+        rectRightX = 44
+        rectLength = 20
+ 
+        # Calculate where the steps should be using left justification
+        if self.step <= 9:
+            stepLeftX = 86
+        elif self.step > 9 and self.step <= 99:
+            stepLeftX = 78
+        else:
+            stepLeftX = 70
+
+        # current step
+        oled.text(str(self.step) + '|' + str(self.patternLength), stepLeftX, 0, 1)
+
+        # Upper + lower values
+        oled.text(str(self.upper), 0, 6, 1)
+        oled.rect(0 , 17, 16, 1, 1)
+        oled.text(str(self.lower), 0, 22, 1)
+
+        # probabilities
+        oled.rect(rectLeftX, 6, rectLength, 8, 1)
+        oled.fill_rect(rectLeftX, 6, self.upperProb1//5, 8, 1)
+        oled.rect(rectRightX, 6, rectLength, 8, 1)
+        oled.fill_rect(rectRightX, 6, self.upperProb2//5, 8, 1)
+        oled.rect(rectLeftX, 22, rectLength, 8, 1)
+        oled.fill_rect(rectLeftX, 22, self.lowerProb1//5, 8, 1)
+        oled.rect(rectRightX, 22, rectLength, 8, 1)
+        oled.fill_rect(rectRightX, 22, self.lowerProb2//5, 8, 1)
+
+        if self.doubleTimeManualOverride or self.doubleTime:
+            oled.text('!¿', 100, 22, 1)
```

Comment:
[question] Do we really have this character available? `¿`

Reply:
No we don't. I changed this to !! instead.